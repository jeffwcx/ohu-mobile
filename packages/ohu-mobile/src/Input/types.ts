import { IconProperty } from '../types';
import { SyntheticEvent, InputHTMLAttributes } from 'vue-tsx-support/types/dom';

export interface InputBaseAttrs {
  max?: number | string;
  min?: number | string;
  maxlength?: string | number;
  minlength?: string | number;
  multiple?: boolean;
  tabindex?: string | number;
  step?: string | number;
  size?: string | number;
  required?: boolean;
  pattern?: string;
  accept?: string | number;
  autocomplete?: string;
  autofocus?: boolean;
  autosave?: string;
  placeholder?: string;
  rows?: string | number;
  cols?: string | number;
}

export interface InputProps extends InputBaseAttrs {
  name?: string;
  type?: string;
  value?: any;
  // show clear button?
  allowClear?: boolean;
  // show toggle eye?
  allowTogglePassword?: boolean;
  disabled?: boolean;
  readonly?: boolean;
  outline?: boolean;
  startAdornment?: IconProperty | string;
  endAdornments?: IconProperty | string;
  noBorder?: boolean;
}

export type InputChangeEvent = SyntheticEvent<
  HTMLInputElement & { composing: boolean }
>;
export type InputFocusEvent = SyntheticEvent<InputHTMLAttributes, FocusEvent>;
export type InputBlurEvent = InputFocusEvent;

export interface InputEvents {
  onValueChange: any;
  onInput: InputChangeEvent;
  onChange: InputChangeEvent;
  onFocus: InputFocusEvent;
  onBlur: InputBlurEvent;
  onClear: void;
  onKeyDown: Event;
  onEnter: Event;
}
