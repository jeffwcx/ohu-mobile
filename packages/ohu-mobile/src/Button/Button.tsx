import { VNodeData } from 'vue';
import Icon from '../Icon';
import { IconDef } from '../types';
import { ButtonEvents, ButtonProps, ButtonTypes, ButtonSizes } from './types';
import { LoaderTailOutlined } from '@ohu-mobile/icons';
import { defineComponent, props } from '../_utils/defineComponent';
import navigate from '../_utils/navigate';

export const buttonProps = {
  htmlType: props(String).default('button'),
  type: props.ofType<ButtonTypes>().default('default'),
  plain: props(Boolean).default(false),
  size: props.ofType<ButtonSizes>().default('lg'),
  loading: props(Boolean).default(false),
  disabled: props(Boolean).default(false),
  inline: props(Boolean).default(false),
  icon: props<string, IconDef>(String, Object).optional,
  round: props(Boolean).default(false),
  link: props(Boolean).default(false),
  tabindex: props(Number).default(0),
  to: props<string, Location>(String, Object).optional,
  url: String,
  replace: props(Boolean).default(false),
};

const Button = defineComponent<ButtonProps, ButtonEvents>('btn').create({
  props: buttonProps,
  methods: {
    onClick(e: Event) {
      if (!this.disabled) {
        navigate(this);
        this.$emit('click', e);
      }
    },
    onTouchstart(e: TouchEvent) {
      this.$emit('touchstart', e);
    },
  },
  render() {
    let { disabled, icon, loading, tabindex, $attrs, $slots } = this;
    if (loading) {
      icon = LoaderTailOutlined;
    }
    const root = this.$rootCls();
    root.is([
      this.round && 'round',
      this.type,
      this.link && 'link',
      this.size,
      this.inline && 'inline',
      !this.inline && this.size === 'lg' && 'block',
      this.plain && 'plain',
      this.loading && 'loading',
      !this.$slots.default && !!this.icon && 'icon-only',
    ]);
    const buttonProps: VNodeData = {
      attrs: {
        ...$attrs,
        role: 'button',
        disabled,
        tabindex,
        // fix 'Form submission canceled because the form is not connected'
        type: this.htmlType,
      },
      on: {
        click: this.onClick,
        touchstart: this.onTouchstart,
      },
      class: root,
    };
    return (
      <button {...buttonProps}>
        {icon && (
          <i>
            <Icon type={icon} spin={loading}></Icon>
          </i>
        )}
        {icon ? (
          $slots.default ? (
            <span>{$slots.default}</span>
          ) : (
            ''
          )
        ) : (
          $slots.default
        )}
      </button>
    );
  },
});

export default Button;
