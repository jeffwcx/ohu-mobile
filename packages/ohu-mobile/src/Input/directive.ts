import { VNode } from 'vue';

interface InputBaseElement extends Element {
  vmodel?: boolean;
  type?: string;
}
export const inBrowser = typeof window !== 'undefined';
export const UA = inBrowser && window.navigator.userAgent.toLowerCase();
export const isIE9 = UA && UA.indexOf('msie 9.0') > 0;
function makeMap(str: string, expectsLowerCase?: boolean) {
  const map = Object.create(null);
  const list = str.split(',');
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < list.length; i += 1) {
    map[list[i]] = true;
  }
  return expectsLowerCase
    ? (val: string) => map[val.toLowerCase()]
    : (val: string) => map[val];
}
const isTextInputType = makeMap('text,number,password,search,email,tel,url');

function onCompositionStart(e: any) {
  e.target.composing = true;
}

function onCompositionEnd(e: any) {
  // prevent triggering an input event for no reason
  if (!e.target.composing) {
    return;
  }
  e.target.composing = false;
  trigger(e.target, 'input');
}

function trigger(el: InputBaseElement, type: string) {
  const e = document.createEvent('HTMLEvents');
  e.initEvent(type, true, true);
  el.dispatchEvent(e);
}

/* istanbul ignore if */
if (isIE9) {
  // http://www.matts411.com/post/internet-explorer-9-oninput/
  document.addEventListener('selectionchange', () => {
    const el = document.activeElement as InputBaseElement;
    if (el && el.vmodel) {
      trigger(el, 'input');
    }
  });
}

export default {
  inserted(el: InputBaseElement, binding: any, vnode: VNode) {
    if (vnode.tag === 'textarea' || isTextInputType(el.type || '')) {
      if (!binding.modifiers || !binding.modifiers.lazy) {
        el.addEventListener('compositionstart', onCompositionStart);
        el.addEventListener('compositionend', onCompositionEnd);
        // Safari < 10.2 & UIWebView doesn't fire compositionend when
        // switching focus before confirming composition choice
        // this also fixes the issue where some browsers e.g. iOS Chrome
        // fires "change" instead of "input" on autocomplete.
        el.addEventListener('change', onCompositionEnd);
        /* istanbul ignore if */
        if (isIE9) {
          el.vmodel = true;
        }
      }
    }
  },
};
