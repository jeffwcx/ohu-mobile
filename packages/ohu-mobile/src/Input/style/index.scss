@use '../../_styles/variables' as *;
@use "pkg:@ohu-mobile/styles" as *;
@use '../../_styles/mixins' as *;

$inputPrefixClass: #{$prefix}input;

.#{$inputPrefixClass} {
  box-sizing: border-box;
  font-size: $inputFontSize;
  font-weight: normal;
  color: $inputColor;
  border: 0;
  display: inline-flex;
  align-items: center;
  &.has-border.is-normal {
    @include hair-border($inputBorderColor, all, $inputBorderRadius);
  }
  &.has-border.is-outline {
    border-style: solid;
    @include hair-border($inputBorderColor, bottom);
  }
  &.is-normal {
    padding: $inputPadding;
    border-radius: $inputBorderRadius;
    &.is-focus {
      &::after {
        border: 2Px solid $inputFocusBorderColor;
      }
    }
    &.has-start-adornment {
      padding: $inputPaddingWithAdornment;
    }
  }
  &.is-outline {
    padding: $inputOutlinePadding;
    &.is-focus {
      &::after {
        border: 1Px solid $inputFocusBorderColor;
      }
    }
    &.has-start-adornment {
      padding: $inputOutlinePaddingWithAdornment;
    }
    &.is-disabled {
      border-style: dashed;
    }
  }
  &.is-disabled {
    cursor: not-allowed;
    opacity: .4;
    input[disabled] {
      cursor: not-allowed;
    }
  }
  &__base {
    box-sizing: border-box;
    font: inherit;
    border: 0;
    margin: 0;
    padding: 0;
    background: none;
    display: block;
    width: 100%;
    color: $inputColor;
    -webkit-tap-highlight-color: transparent;
    &::placeholder {
      color: $inputPlaceholderColor;
    }
    &:focus {
      outline: 0;
    }
    & ~ .#{$inputPrefixClass}__adornment {
      margin-left: $inputAdornmentMargin;
    }
  }
  input {
    line-height: $inputLineHeight;
  }
  textarea {
    height: auto;
    resize: none;
    line-height: $inputTextAreaLineHeight;
  }
  &__adornment {
    flex-shrink: 0;
    width: 1.5em;
    height: 1.5em;
    color: $inputAdornmentColor;
    display: flex;
    align-items: center;
    justify-content: center;
    white-space: nowrap;
    outline: 0;
    > .#{$prefix}icon {
      font-size: 1.125em;
    }
    span {
      font-size: $inputAdornmentFontSize;
    }
    > .#{$prefix}btn.is-link {
      flex-shrink: 0;
      font-size: 1em;
      color: $inputAdornmentColor;
      i {
        font-size: 1.125em;
      }
    }
    & ~ .#{$inputPrefixClass}__base {
      margin-left: $inputAdornmentMargin;
    }
    &.is-auto {
      width: auto;
    }
  }
}
