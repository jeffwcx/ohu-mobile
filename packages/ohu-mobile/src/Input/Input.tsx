import { defineComponent, props } from '../_utils/defineComponent';
import {
  InputProps,
  InputEvents,
  InputBlurEvent,
  InputFocusEvent,
  InputChangeEvent,
} from './types';
import directive from './directive';
import { IconDef } from '../types';
import { IconProps } from '../Icon';
import { getIcon } from '../_utils/icon-utils';
import Button from '../Button';
import {
  CloseCircleFilled,
  EyeCloseOutlined,
  EyeOutlined,
} from '@ohu-mobile/icons';
import { VNodeData, VNode } from 'vue/types/umd';
import { BlockContext } from '../_utils/classHelper';
import { fieldMixin } from '../Form/fieldMixin';
import { FormFieldInnerMethods } from '../Form/types';

export interface InputMethods extends FormFieldInnerMethods {
  stateValue: any;
  showPassword: boolean;
  isInputFocus: boolean;
  // renderInput: () => JSX.Element;
  renderTextArea: () => JSX.Element;
  getAdornment: (adornment?: IconDef | IconProps | string) => VNode | undefined;
  handleFocus: (e: InputFocusEvent) => void;
  handleBlur: (e: InputFocusEvent) => void;
  handleChange: (e: InputChangeEvent) => void;
  handleKeyDown: (e: KeyboardEvent) => void;
  togglePassword: (e: Event) => void;
  clear: (e: Event) => void;
}

const Input = defineComponent<InputProps, InputEvents, {}, InputMethods>(
  'input',
)
  .mixin(fieldMixin('stateValue', 'value'))
  .create({
    model: {
      prop: 'value',
      event: 'valueChange',
    },
    props: {
      name: props(String).optional,
      type: props(String).default('text'),
      outline: props(Boolean).default(false),
      value: props.ofAny().optional,
      allowClear: props(Boolean).default(false),
      allowTogglePassword: props(Boolean).default(true),
      noBorder: props(Boolean).default(false),
      disabled: props(Boolean).default(false),
      readonly: props(Boolean).default(false),
      placeholder: props(String).default(''),
      startAdornment: props<IconDef, IconProps, string>(Object, Object, String)
        .optional,
      endAdornment: props<IconDef, IconProps, string>(Object, Object, String)
        .optional,
      max: props(String, Number).optional,
      min: props(String, Number).optional,
      maxlength: props(String, Number).optional,
      minlength: props(String, Number).optional,
      multiple: props(Boolean).optional,
      tabindex: props(String, Number).optional,
      step: props(String, Number).optional,
      size: props(String, Number).optional,
      required: props(Boolean).default(false),
      pattern: props(String).optional,
      accept: props(String, Number).optional,
      autocomplete: props(String).optional,
      autofocus: props(Boolean).default(false),
      autosave: props(String).optional,
      rows: props(String, Number).optional,
      cols: props(String, Number).optional,
    },
    directives: {
      inputBase: directive,
    },
    watch: {
      value(cur: any) {
        this.stateValue = cur;
      },
    },
    data() {
      return {
        showPassword: false,
        isInputFocus: false,
      };
    },
    methods: {
      togglePassword(e: Event) {
        e.stopPropagation();
        this.showPassword = !this.showPassword;
      },
      clear(e: Event) {
        e.stopPropagation();
        const input = this.$refs.input as HTMLInputElement;
        if (input) {
          input.value = '';
        }
        this.stateValue = '';
        this.$emit('valueChange', this.stateValue);
        this.$emit('change', this.stateValue);
        this.$emit('clear');
      },
      focus() {
        const input = this.$refs.input as HTMLInputElement;
        input && input.focus();
      },
      blur() {
        const input = this.$refs.input as HTMLInputElement;
        input && input.blur();
      },
      handleBlur(e: InputBlurEvent) {
        this.isInputFocus = false;
        this.$emit('blur', e);
      },
      handleFocus(e: InputFocusEvent) {
        this.isInputFocus = true;
        this.$emit('focus', e);
      },
      handleChange(e: InputChangeEvent) {
        const target = e.target;
        if (!target) return;
        this.stateValue = target.value;
        if (!target.composing) {
          this.$emit('valueChange', this.stateValue);
        }
        this.$emit('change', e);
        this.$emit('input', e);
      },
      handleKeyDown(e: KeyboardEvent) {
        this.$emit('keydown', e);
        if (e.keyCode === 13) {
          e.preventDefault();
          this.$emit('enter', e);
        }
      },
      getAdornment(adornment?: IconDef | IconProps | string) {
        if (!adornment) return;
        if (typeof adornment === 'string') {
          return <span>{adornment}</span>;
        } else {
          return getIcon(this.$createElement, adornment);
        }
      },
      renderInput(root: BlockContext, attrs: Record<string, any>) {
        let {
          type,
          allowClear,
          allowTogglePassword,
          startAdornment,
          endAdornment,
        } = this;
        let start = this.getAdornment(startAdornment);
        if (start) {
          start = <div class={root.element('adornment')}>{start}</div>;
        }
        let end;
        if (this.$slots.endAdornment) {
          end = (
            <div class={root.element('adornment').is('auto')}>
              {this.$slots.endAdornment}
            </div>
          );
        } else {
          end = this.getAdornment(endAdornment);
          if (end) {
            end = <div class={root.element('adornment')}>{end}</div>;
          }
        }
        const inputType = this.showPassword ? 'text' : type;
        const inputProps: VNodeData = {
          domProps: {
            value: this.stateValue,
            type: inputType,
          },
          class: root.element('base'),
          attrs,
          on: {
            ...this.$listeners,
            focus: this.handleFocus,
            blur: this.handleBlur,
            input: this.handleChange,
            keydown: this.handleKeyDown,
          },
          directives: [{ name: 'inputBase' }],
          ref: 'input',
        };
        if (this.type === 'textarea') {
          return <textarea {...inputProps}></textarea>;
        }
        return [
          start,
          <input {...inputProps} />,
          allowClear && !this.readonly && !this.disabled && (
            <div class={root.element('adornment').is('button')} tabindex={-1}>
              {this.stateValue && (
                <Button
                  tabindex={-1}
                  link
                  size="md"
                  round
                  icon={CloseCircleFilled}
                  onClick={this.clear}
                />
              )}
            </div>
          ),
          end,
          type === 'password' && allowTogglePassword && (
            <div class={root.element('adornment').is('button')} tabindex={-1}>
              <Button
                tabindex={-1}
                link
                size="md"
                round
                icon={this.showPassword ? EyeOutlined : EyeCloseOutlined}
                onClick={this.togglePassword}
              />
            </div>
          ),
        ];
      },
    },
    mounted() {
      if (this.autofocus) {
        this.focus();
      }
    },
    render() {
      const root = this.$rootCls();
      let {
        type,
        value,
        outline,
        allowClear,
        allowTogglePassword,
        startAdornment,
        endAdornment,
        noBorder,
        autofocus,
        name,
        ...attrs
      } = this.$props;
      if (attrs) {
        attrs.name = this.fieldName || name;
      }
      return (
        <div
          class={root
            .has([
              this.startAdornment && 'start-adornment',
              !this.noBorder && 'border',
            ])
            .is([
              this.isInputFocus && 'focus',
              this.outline ? 'outline' : 'normal',
              this.disabled && 'disabled',
            ])}
        >
          {this.renderInput(root, attrs)}
        </div>
      );
    },
  });

export default Input;
