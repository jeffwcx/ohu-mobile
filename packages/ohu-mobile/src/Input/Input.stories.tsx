import Vue from 'vue';

import Card from '../Card';
import '../Card/style';
import Input from './index';
import './style';
import { EditOutlined } from '@ohu-mobile/icons';
import Button from '../Button';
import '../Button/style';

export default {
  title: 'Components/Form/Input',
  parameters: {
    component: Input,
    options: {
      showPanel: true,
    },
  },
};

export const Basic = () =>
  Vue.extend({
    data() {
      return {
        basic: 'hello',
        adornment: '',
        password: '',
      };
    },
    render() {
      return (
        <div>
          <Card shadow divider>
            <Card.Header>Basic</Card.Header>
            <Input v-model={this.basic} allowClear placeholder="please input" />
            <p>{this.basic}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Readonly</Card.Header>
            <Input readonly v-model={this.basic} allowClear />
            <p>{this.basic}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Disabled</Card.Header>
            <Input disabled v-model={this.basic} allowClear />
            <p>{this.basic}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Adornment</Card.Header>
            <Input
              allowClear
              v-model={this.adornment}
              startAdornment={EditOutlined}
              endAdornment="kg"
            />
            <p>{this.adornment}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>ButtonAdornment</Card.Header>
            <Input
              v-model={this.adornment}
              allowClear
              placeholder="请输入验证码"
              startAdornment={EditOutlined}
            >
              <Button slot="endAdornment" type="primary" size="sm">
                获取验证码
              </Button>
            </Input>
            <p>{this.adornment}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Password</Card.Header>
            <Input
              v-model={this.password}
              type="password"
              allowClear
              style="width: 100%"
              placeholder="请输入密码"
            />
            <p>{this.password}</p>
          </Card>
        </div>
      );
    },
  });

export const Outline = () =>
  Vue.extend({
    data() {
      return {
        basic: '',
        adornment: '',
        password: '',
      };
    },
    render() {
      return (
        <div>
          <Card shadow divider>
            <Card.Header>Basic</Card.Header>
            <Input
              outline
              v-model={this.basic}
              allowClear
              placeholder="please input"
            />
            <p>{this.basic}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Readonly</Card.Header>
            <Input outline readonly v-model={this.basic} allowClear />
            <p>{this.basic}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Disabled</Card.Header>
            <Input outline disabled v-model={this.basic} allowClear />
            <p>{this.basic}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Adornment</Card.Header>
            <Input
              outline
              v-model={this.adornment}
              startAdornment={EditOutlined}
              endAdornment="kg"
            />
            <p>{this.adornment}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>ButtonAdornment</Card.Header>
            <Input
              v-model={this.adornment}
              outline
              startAdornment={EditOutlined}
            >
              <Button slot="endAdornment" type="primary" size="sm">
                获取验证码
              </Button>
            </Input>
            <p>{this.adornment}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Password</Card.Header>
            <Input
              outline
              v-model={this.password}
              type="password"
              allowClear
              style="width: 100%"
              placeholder="请输入密码"
            />
            <p>{this.password}</p>
          </Card>
        </div>
      );
    },
  });

export const Textarea = () =>
  Vue.extend({
    data() {
      return {
        value: '',
      };
    },
    render() {
      return (
        <div>
          <Card shadow divider>
            <Card.Header>Basic</Card.Header>
            <Input
              v-model={this.value}
              placeholder="请输入"
              type="textarea"
              rows={3}
              cols={30}
            />
            <p>{this.value}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Readonly</Card.Header>
            <Input
              v-model={this.value}
              readonly
              type="textarea"
              rows={3}
              cols={30}
            />
            <p>{this.value}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>Disabled</Card.Header>
            <Input
              v-model={this.value}
              disabled
              readonly
              type="textarea"
              rows={3}
              cols={30}
            />
            <p>{this.value}</p>
          </Card>
          <Card shadow divider>
            <Card.Header>outline</Card.Header>
            <Input
              v-model={this.value}
              outline
              type="textarea"
              rows={3}
              cols={30}
            />
            <p>{this.value}</p>
          </Card>
        </div>
      );
    },
  });
