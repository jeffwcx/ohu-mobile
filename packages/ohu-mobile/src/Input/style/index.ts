import '../../_styles';
import './index.scss';

// style dependencies
import '../../Button/style';
