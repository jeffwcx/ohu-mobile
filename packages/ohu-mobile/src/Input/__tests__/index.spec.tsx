import { mount } from '@vue/test-utils';
import Vue from 'vue';
import Input from '..';
import EditBoxFilled from '@ohu-mobile/icons/lib/EditBoxFilled';
import Button from '../../Button';
import { describe, it, expect, vi } from 'vitest';

function createDiv() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}

describe('Input', () => {
  it('basic', () => {
    const wrapper = mount(Input, {});
    expect(wrapper.element).toMatchSnapshot();
  });
  it('v-model', async () => {
    const Test = Vue.extend({
      data() {
        return {
          value: '1',
        };
      },
      render() {
        return <Input v-model={this.value} />;
      },
    });
    const wrapper = mount(Test);
    await wrapper.setData({
      value: '2',
    });
    const input = wrapper.find('input').element as HTMLInputElement;
    expect(input.value).toEqual('2');
    input.value = '1';
    await wrapper.find('input').trigger('input');
    expect(wrapper.vm.value).toEqual('1');
  });
  it('allowClear', () => {
    const func = vi.fn();
    const Test = Vue.extend({
      data() {
        return {
          value: '1',
        };
      },
      render() {
        return <Input v-model={this.value} allowClear onClear={func} />;
      },
    });
    const wrapper = mount(Test);
    wrapper.vm.value = '2';
    expect(wrapper.element).toMatchSnapshot();
    wrapper.find('.ohu-input__adornment button').trigger('click');
    setTimeout(() => {
      expect(wrapper.vm.value).toEqual('');
      expect(func).toBeCalled();
    });
  });

  it('allowTogglePassword', () => {
    const Test = Vue.extend({
      render() {
        return <Input type="password" />;
      },
    });
    const wrapper = mount(Test);
    expect(wrapper.element).toMatchSnapshot();
    const input = wrapper.find('input').element as HTMLInputElement;
    input.value = 'password';
    wrapper.find('button').trigger('click');
    Vue.nextTick(() => {
      expect(input.type).toBe('text');
      wrapper.find('button').trigger('click');
      Vue.nextTick(() => {
        expect(wrapper.element).toMatchSnapshot();
        expect(input.type).toBe('password');
      });
    });
  });

  it('outline', () => {
    const wrapper = mount(Input, {
      propsData: {
        outline: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('noBorder', () => {
    const wrapper = mount(Input, {
      propsData: {
        noBorder: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('adornment', () => {
    const wrapper = mount(Input, {
      propsData: {
        startAdornment: 'kg',
        endAdornment: EditBoxFilled,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('buttonAdornment', () => {
    const wrapper = mount(Input, {
      slots: {
        endAdornment: Button,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('textarea', () => {
    const wrapper = mount(Input, {
      propsData: {
        type: 'textarea',
        rows: 2,
        cols: 4,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('disabled and readony', () => {
    const wrapper = mount(Input, {
      propsData: {
        disabled: true,
        readonly: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('focus and blur event', async () => {
    const blurFunc = vi.fn();
    const focusFunc = vi.fn();
    const wrapper = mount(Input, {
      propsData: {
        type: 'text',
      },
      listeners: {
        blur: blurFunc,
        focus: focusFunc,
      },
      attachTo: createDiv(),
    });
    await wrapper.find('input').trigger('focus');
    expect(focusFunc).toBeCalled();
    await wrapper.find('input').trigger('blur');
    expect(blurFunc).toBeCalled();
  });

  it('enter event', async () => {
    const func = vi.fn();
    const Test = Vue.extend({
      render() {
        return <Input onEnter={func} />;
      },
    });
    const wrapper = mount(Test);
    await wrapper.find('input').trigger('keydown.enter');
    expect(func).toBeCalled();
  });

  it('event valueChange and change', async () => {
    const valueChangeFunc = vi.fn();
    const changeFunc = vi.fn();
    const wrapper = mount(Input, {
      listeners: {
        valueChange: valueChangeFunc,
        change: changeFunc,
      },
      attachTo: createDiv(),
    });
    const input = wrapper.find('input').element as HTMLInputElement;
    input.value = 'value';
    await wrapper.find('input').trigger('input');
    expect(valueChangeFunc).toBeCalled();
    expect(changeFunc).toBeCalled();
  });
});
