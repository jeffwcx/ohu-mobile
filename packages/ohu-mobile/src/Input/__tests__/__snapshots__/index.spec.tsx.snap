// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > adornment 1`] = `
<div
  class="ohu-input has-start-adornment has-border is-normal"
>
  <div
    class="ohu-input__adornment"
  >
    <span>
      kg
    </span>
  </div>
  <input
    class="ohu-input__base"
    placeholder=""
    type="text"
  />
  <div
    class="ohu-input__adornment"
  >
    <svg
      aria-hidden="true"
      class="ohu-icon"
      focusable="false"
      role="presentation"
      tabindex="-1"
      title="edit-box"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="m16.757 3-7.466 7.466.008 4.247 4.238-.007L21 7.243V20a1 1 0 0 1-1 1H4a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1zm3.728-.9L21.9 3.516l-9.192 9.192-1.412.003-.002-1.417z"
      />
    </svg>
  </div>
</div>
`;

exports[`Input > allowClear 1`] = `
<div
  class="ohu-input has-border is-normal"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="text"
  />
  <div
    class="ohu-input__adornment is-button"
    tabindex="-1"
  >
    <button
      class="ohu-btn is-round is-default is-link is-md is-icon-only"
      role="button"
      tabindex="-1"
      type="button"
    >
      <i>
        <svg
          aria-hidden="true"
          class="ohu-icon"
          focusable="false"
          role="presentation"
          tabindex="-1"
          title="close-circle"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10m0-11.414L9.172 7.757 7.757 9.172 10.586 12l-2.829 2.828 1.415 1.415L12 13.414l2.828 2.829 1.415-1.415L13.414 12l2.829-2.828-1.415-1.415z"
          />
        </svg>
      </i>
    </button>
  </div>
</div>
`;

exports[`Input > allowTogglePassword 1`] = `
<div
  class="ohu-input has-border is-normal"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="password"
  />
  <div
    class="ohu-input__adornment is-button"
    tabindex="-1"
  >
    <button
      class="ohu-btn is-round is-default is-link is-md is-icon-only"
      role="button"
      tabindex="-1"
      type="button"
    >
      <i>
        <svg
          aria-hidden="true"
          class="ohu-icon"
          focusable="false"
          role="presentation"
          tabindex="-1"
          title="eye-close"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="m9.342 18.782-1.931-.518.787-2.939a11 11 0 0 1-3.237-1.872l-2.153 2.154-1.415-1.415 2.154-2.153a10.96 10.96 0 0 1-2.371-5.07l1.968-.359C3.903 10.812 7.579 14 12 14c4.42 0 8.097-3.188 8.856-7.39l1.968.358a10.96 10.96 0 0 1-2.37 5.071l2.153 2.153-1.415 1.415-2.153-2.154a11 11 0 0 1-3.237 1.872l.787 2.94-1.931.517-.788-2.94a11.1 11.1 0 0 1-3.74 0z"
          />
        </svg>
      </i>
    </button>
  </div>
</div>
`;

exports[`Input > allowTogglePassword 2`] = `
<div
  class="ohu-input has-border is-normal"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="password"
  />
  <div
    class="ohu-input__adornment is-button"
    tabindex="-1"
  >
    <button
      class="ohu-btn is-round is-default is-link is-md is-icon-only"
      role="button"
      tabindex="-1"
      type="button"
    >
      <i>
        <svg
          aria-hidden="true"
          class="ohu-icon"
          focusable="false"
          role="presentation"
          tabindex="-1"
          title="eye-close"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="m9.342 18.782-1.931-.518.787-2.939a11 11 0 0 1-3.237-1.872l-2.153 2.154-1.415-1.415 2.154-2.153a10.96 10.96 0 0 1-2.371-5.07l1.968-.359C3.903 10.812 7.579 14 12 14c4.42 0 8.097-3.188 8.856-7.39l1.968.358a10.96 10.96 0 0 1-2.37 5.071l2.153 2.153-1.415 1.415-2.153-2.154a11 11 0 0 1-3.237 1.872l.787 2.94-1.931.517-.788-2.94a11.1 11.1 0 0 1-3.74 0z"
          />
        </svg>
      </i>
    </button>
  </div>
</div>
`;

exports[`Input > basic 1`] = `
<div
  class="ohu-input has-border is-normal"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="text"
  />
</div>
`;

exports[`Input > buttonAdornment 1`] = `
<div
  class="ohu-input has-border is-normal"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="text"
  />
  <div
    class="ohu-input__adornment is-auto"
  >
    <button
      class="ohu-btn is-default is-lg is-block"
      role="button"
      tabindex="0"
      type="button"
    />
  </div>
</div>
`;

exports[`Input > disabled and readony 1`] = `
<div
  class="ohu-input has-border is-normal is-disabled"
>
  <input
    class="ohu-input__base"
    disabled="disabled"
    placeholder=""
    readonly="readonly"
    type="text"
  />
</div>
`;

exports[`Input > noBorder 1`] = `
<div
  class="ohu-input is-normal"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="text"
  />
</div>
`;

exports[`Input > outline 1`] = `
<div
  class="ohu-input has-border is-outline"
>
  <input
    class="ohu-input__base"
    placeholder=""
    type="text"
  />
</div>
`;

exports[`Input > textarea 1`] = `
<div
  class="ohu-input has-border is-normal"
>
  <textarea
    class="ohu-input__base"
    cols="4"
    placeholder=""
    rows="2"
  />
</div>
`;
